receiver/sqlserverreceiver/scraper.go

```diff
@@ -299,3 +326,243 @@ func (s *sqlServerScraperHelper) recordDatabaseStatusMetrics(ctx context.Context
+			record.Attributes().PutInt(dbPrefix+physicalReads, diff)
+		}
+
+		// handling `execution_count`
+		totalExecutionCount, err := strconv.ParseInt(row[executionCount], 10, 64)
+		if err != nil {
+			err = fmt.Errorf("row %d: %w", i, err)
+			errs = append(errs, err)
+		} else {
+			if cached, diff := s.cacheAndDiff(queryHashVal, queryPlanHashVal, executionCount, totalExecutionCount); cached {
+				record.Attributes().PutInt(dbPrefix+executionCount, diff)
+			}
+		}
+
+		// handle `total_worker_time`, storing milliseconds
+		workerTime, err := strconv.ParseInt(row[totalWorkerTime], 10, 64)
+		if err != nil {
+			err = fmt.Errorf("row %d: %w", i, err)
+			errs = append(errs, err)
+		} else {
+			if cached, diff := s.cacheAndDiff(queryHashVal, queryPlanHashVal, totalWorkerTime, workerTime/1000); cached {
+				record.Attributes().PutInt(dbPrefix+totalWorkerTime, diff)
+			}
+		}
+
+		// handle `total_grant_kb`
+		memoryGranted, err := strconv.ParseInt(row[totalGrant], 10, 64)
+		if err != nil {
+			err = fmt.Errorf("row %d: %w", i, err)
+			errs = append(errs, err)
+		} else {
+			if cached, diff := s.cacheAndDiff(queryHashVal, queryPlanHashVal, totalGrant, memoryGranted); cached {
+				record.Attributes().PutInt(dbPrefix+totalGrant, diff)
+			}
+		}
+
+		// handling `query_text`
+		obfuscatedSQL, err := obfuscateSQL(row[queryText])
+		if err != nil {
+			err = fmt.Errorf("row %d: %w", i, err)
+			errs = append(errs, err)
+		}
+		record.Attributes().PutStr(dbPrefix+queryText, obfuscatedSQL)
+
+		// handling `query_plan`
+		obfuscatedQueryPlan, err := obfuscateXMLPlan(row[queryPlan])
+		if err != nil {
+			err = fmt.Errorf("row %d: %w", i, err)
+			errs = append(errs, err)
+		}
+		record.Attributes().PutStr(dbPrefix+queryPlan, obfuscatedQueryPlan)
+	}
+
+	return logs, errors.Join(errs...)
+}
+
+// cacheAndDiff store row(in int) with query hash and query plan hash variables
+// (1) returns true if the key is cached before
+// (2) returns positive value if the value is larger than the cached value
+func (s *sqlServerScraperHelper) cacheAndDiff(queryHash string, queryPlanHash string, column string, val int64) (bool, int64) {
+	if s.cache == nil {
+		s.logger.Error("LRU cache is not successfully initialized, skipping caching and diffing")
+		return false, 0
+	}
+
+	if val < 0 {
+		return false, 0
+	}
+
+	key := queryHash + "-" + queryPlanHash + "-" + column
+
+	cached, ok := s.cache.Get(key)
+	if !ok {
+		s.cache.Add(key, val)
+		return false, val
+	}
+
+	if val > cached {
+		s.cache.Add(key, val)
+		return true, val - cached
+	}
+
+	return true, 0
+}
+
+// sortRows sorts the rows based on the `values` slice in descending order and return the first M(M=maximum) rows
+// Input: (row: [row1, row2, row3], values: [100, 10, 1000], maximum: 2
+// Expected Output: (row: [row3, row1]
+func sortRows(rows []sqlquery.StringMap, values []int64, maximum uint) []sqlquery.StringMap {
```

Comment:
I feel like this would be better using a heap or tree structure instead of an array, then you can have an iterator that performs path walking algorithm to get the data you need.

Reply:
improved to use a priority queue